### PR TITLE
List chains

### PR DESCRIPTION
**Pull request by jniewt**
#21 Implemented basic functionality. All chains from all tables are being listed. This is my first contribution so let me know if I can improve something.

### CODE REVIEW

**Comment by jniewt** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
>  **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fgoogle%2Fnftables%2Fpull%2F25) for more info**.

I signed it.

File: chain.go [hunk 1]
```diff
@@ -106,3 +110,94 @@ func (cc *Conn) AddChain(c *Chain) *Chain {
...
+			})
+		}
+
+	}
+
+	return &c, nil
+}
+
+func hookFromMsg(b []byte) (ChainHook, ChainPriority, error) {
+
```
Comment (stapelberg) [inline chain.go]: nit: remove blank line

File: chain.go [hunk 2]
```diff
@@ -106,3 +110,94 @@ func (cc *Conn) AddChain(c *Chain) *Chain {
...
+}
+
+func chainFromMsg(msg netlink.Message) (*Chain, error) {
+	if got, want := msg.Header.Type, chainHeaderType; got != want {
+		return nil, fmt.Errorf("unexpected header type: got %v, want %v", got, want)
+	}
+
+	var c Chain
+
+	ad, err := netlink.NewAttributeDecoder(msg.Data[4:])
```
Comment (stapelberg) [inline chain.go]: What are you skipping with `msg.Data[4:]`? Is this always safe, or could msg.Data be shorter than 4?

This deserves a comment in the code.

File: chain.go [hunk 3]
```diff
@@ -15,13 +15,17 @@
...
+	"encoding/binary"
+	"fmt"
 	"math"
 
 	"github.com/google/nftables/binaryutil"
 	"github.com/mdlayher/netlink"
 	"golang.org/x/sys/unix"
 )
 
+var chainHeaderType = netlink.HeaderType((unix.NFNL_SUBSYS_NFTABLES << 8) | unix.NFT_MSG_NEWCHAIN)
```
Comment (stapelberg) [inline chain.go]: Why is this variable defined at package-level? Can you move it into chainFromMsg please?

File: chain.go [hunk 4]
```diff
@@ -106,3 +110,94 @@ func (cc *Conn) AddChain(c *Chain) *Chain {
...
 	return c
 }
+
+// ListChains returns currently configured chains in the kernel
+func (cc *Conn) ListChains() ([]*Chain, error) {
+	conn, err := cc.dialNetlink()
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (stapelberg) [inline chain.go]: nit: remove blank line

File: chain.go (hunk 2 above)
Comment (jniewt) [reply chain.go]: I couldn't find out, sorry. It's this way in tables.go too.

**Comment by stapelberg** [comment]
Thanks!

File: chain.go (hunk 2 above)
Comment (jniewt) [reply chain.go]: My best guess is that the 4 bytes are NFTL_CHAIN_FAMILY which is not part of nft_chain_attributes so we need to omit them to parse the attributes. No idea why the response would start with a non-attribute though and so couldn't confirm 100%.

**Pull request merged**